### ISSUE TITLE
Modal window or toolwindow for indicating specific progress

### ISSUE THREAD

**Issue by shana** [github/VisualStudio#555]
We need a way of indicating progress of git actions, especially when checking out branches. These actions probably require not allowing the user to do things in VS, since the state of files and solutions is in flux, and VS may possibly want to do git actions itself, which may cause problems.

## The proposal is:

- An interface `IHasProgress`
- A property or observable that triggers on progress state change and reports progress for listeners or
a method with Action callbacks

- A window with:
  - Title with the action currently being undertaken
  - Line of text indicating current progress
  - A generic progress bar to entertain (cats optional)
  - No window buttons

- Showing the window in a modal way without blocking the action being undertaken

- ViewModels that want to report action progress outside of their current view context implement `IHasProgress`.
- The hosting container (for instance, GitHubPaneViewModel) listens for progress notifications and relays that information to the modal window

**Comment by grokys** [comment]
Regarding how to show the progress notification; Team Explorer shows an indeterminate progress strip at the top of the pane - we should probably follow that example. You can see it here above the "Home" label:

![image](https://cloud.githubusercontent.com/assets/1775141/18563384/c38e47ee-7b87-11e6-9080-b296399afe90.png)

**Comment by shana** [comment]
Yes, that was my thinking as well, although sometimes you just don't see it at all, it's really subtle.

**Comment by grokys** [comment]
Though there are two types of progress that we're likely to want to be able to show:

- Operations such as switching branches where you want a modal progress notification, as you don't want someone making changes while it's happening
- Operations such as fetching/reading from the API which shouldn't block the UI

I don't know if you were intending on capturing both those use cases here?

**Comment by shana** [comment]
Ah, no, non-modal busy UI is on #554, busy state issue. The Team Explorer-like progress bar is more for that issue. This issue is for reporting specific progress with real progress information.
